### ISSUE TITLE
Could HTTP2Client talk to HTTP1 Server?

### ISSUE THREAD

**Issue by JaySWang** [eclipse/jetty.project#1308]
Hi there,

HTTP2Client fails to talk to HTTP1 server with the following exception:
```
Caused by: java.io.IOException: 1/unknown_frame_type_80
	at org.eclipse.jetty.http2.HTTP2Session.onConnectionFailure(HTTP2Session.java:471)
	at org.eclipse.jetty.http2.parser.Parser.notifyConnectionFailure(Parser.java:205)
	at org.eclipse.jetty.http2.parser.Parser.parseBody(Parser.java:171)
	at org.eclipse.jetty.http2.parser.Parser.parse(Parser.java:110)
	at org.eclipse.jetty.http2.HTTP2Connection$HTTP2Producer.produce(HTTP2Connection.java:222)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:147)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produce(ExecuteProduceConsume.java:97)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillableBlocking(HTTP2Connection.java:120)
	at org.eclipse.jetty.http2.HTTP2Connection.access$1000(HTTP2Connection.java:41)
	at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:277)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:112)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:129)
	at org.eclipse.jetty.util.thread.strategy.ExecutingExecutionStrategy.invoke(ExecutingExecutionStrategy.java:58)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:201)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:133)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
	... 1 more
```

Could HTTP2Client talk to HTTP1 Server?
Many thanks.

**Comment by sbordet** [comment]
`HTTP2Client` is a HTTP/2 only client.

`HttpClient` is the client that speak by default HTTP/1.1, although it can speak other protocols, see [the documentation](http://www.eclipse.org/jetty/documentation/current/http-client-transport.html).

We have not implemented yet a mechanism, in the client, that can speak different protocols to different domains. We typically suggest to use different `HttpClient` instances configured each with a different transport, since most of the times the client knows what protocol a domain supports.

**Comment by elyograg** [comment]
Not having support for both protocols in one client object is problematic for our attempts to get HTTP/2 support in Solr.  See our issue:

https://issues.apache.org/jira/browse/SOLR-12297

TL;DR info:

HttpSolrClient is the lower-level implementation.  This client talks to one Solr server.  It is immutable in every way, including the http client, once created.

We have LBHttpSolrClient, which is a version that can accept multiple URLs, load balance requests between them, and deal with dead servers.  This client uses HttpSolrClient objects internally.  This implementation is where our difficulties with http version support really begin.  At the moment, this implementation uses one Apache HttpClient object, and that object is shared by all the HttpSolrClient objects used internally.  The natural progression would be to have one jetty http client shared between all the internal objects.

There is also CloudSolrClient, which talks to a SolrCloud cluster via ZooKeeper, dynamically maintains a list of functioning Solr servers in the cloud, and uses that info to configure one LBHttpSolrClient object, already mentioned.

Without the ability to detect the http version on the fly, our client implementations will be required to stay with http/1.1 until an entire cluster has been upgraded.  A client restart would be required to initiate a manual switch to http/2.  We'd really like to be able to do rolling upgrades of servers and have clients automatically start using http/2 without a restart.  If a server has to be downgraded for some reason, it should automatically switch back to http/1.1.

**Comment by joakime** [comment]
@elyograg use the `HttpClient` and the documented pluggable transports to have a single client with support for both HTTP/1.1 and HTTP/2.
Don't use the `HTTP2Client`, as that's protocol specific.

**Comment by sbordet** [comment]
@elyograg the right issue is #1350.

**Comment by lubinson** [comment]
@sbordet 

I also has the similar situation, our servers are dynamic. we don't actually know server is http1.1 or h2c. h2c has a upgrade procedure. If server is http1.1, the upgrade will failed. Can at this situation HttpClient fallback to use http1.1 to talk. 

But h2(https/2) don't have upgrade procedure. So how to use a single one client to handle all http protocols cases. 

Best Thanks.
